### PR TITLE
Display user email for administrators

### LINKED ISSUE #1704: 🚀 [Feature] Make Email of a User visible for Moderators
> [datenbrei](https://github.com/datenbrei) **Authored by [datenbrei](https://github.com/datenbrei)**
_<time datetime="2019-09-24T11:09:45Z" title="Tuesday, September 24th 2019, 1:09:45 pm +02:00">Sep 24, 2019</time>_
_Closed <time datetime="2019-10-07T10:33:34Z" title="Monday, October 7th 2019, 12:33:34 pm +02:00">Oct 7, 2019</time>_
---

## :rocket: Feature

A moderator needs to contact people. For now he can't. Therfore in the admin panels "Users" lookup we need this email address ether (and best) as a small icon which when be clicked it is copied to the clipboard if possible or with a hover.

### Design & Layout

The emails of users should **not** be directly listed, because of privacy.
user-lookup

### Additional context

As long as we do moderation by ourselfes this is ok. Later, if community moderation comes, we need ether an NDA for everybody or we need another method of communication

### PR DESCRIPTION
**Pull request by Tirokk**
> [aonomike](https://github.com/aonomike) **Authored by [aonomike](https://github.com/aonomike)**
_<time datetime="2019-10-03T11:34:45Z" title="Thursday, October 3rd 2019, 1:34:45 pm +02:00">Oct 3, 2019</time>_
_Merged <time datetime="2019-10-07T10:33:34Z" title="Monday, October 7th 2019, 12:33:34 pm +02:00">Oct 7, 2019</time>_
---

## 🍰 Pullrequest

- Display user email to administrators
- Refactor Permission middleware tests
- Add tests to check email of user only revealed to admin and owner

### Issues

fixes #1704 

### Images

Screen Shot 2019-10-05 at 5 29 31 PM

### CODE REVIEW

File: webapp/pages/admin/users.vue
Comment (Tirokk) [inline webapp/pages/admin/users.vue:137]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-03T23:46:33Z" title="Friday, October 4th 2019, 1:46:33 am +02:00">Oct 4, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -128,6 +138,7 @@ export default {
               id
               name
               slug
+              email
```


At the moment, the backend does not allow anybody except the owner to see the `email` of the user. You will need to change permissions in the backend.

I think I would still feel safe, if only the owner *or* an admin user can see the email.

Comment (Tirokk) [reply webapp/pages/admin/users.vue:137]: > [aonomike](https://github.com/aonomike) **Authored by [aonomike](https://github.com/aonomike)**
_<time datetime="2019-10-05T14:24:29Z" title="Saturday, October 5th 2019, 4:24:29 pm +02:00">Oct 5, 2019</time>_

---


Done, thanks @roschaefer

File: backend/src/middleware/permissionsMiddleware.spec.js
Comment (Tirokk) [inline backend/src/middleware/permissionsMiddleware.spec.js:78]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-07T08:58:11Z" title="Monday, October 7th 2019, 10:58:11 am +02:00">Oct 7, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,89 +1,141 @@
-import { GraphQLClient } from 'graphql-request'
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../server'
 import Factory from '../seed/factories'
-import { host, login } from '../jest/helpers'
+import { gql } from '../jest/helpers'
+import { getDriver, neode as getNeode } from '../bootstrap/neo4j'
 
 const factory = Factory()
+const instance = getNeode()
+const driver = getDriver()
+
+let query, authenticatedUser, owner, anotherRegularUser, administrator, variables, moderator
+
+const userQuery = gql`
+  query($name: String) {
+    User(name: $name) {
+      email
+    }
+  }
+`
 
 describe('authorization', () => {
+  beforeAll(async () => {
+    await factory.cleanDatabase()
+    const { server } = createServer({
+      context: () => ({
+        driver,
+        instance,
+        user: authenticatedUser,
+      }),
+    })
+    query = createTestClient(server).query
+  })
+
   describe('given two existing users', () => {
     beforeEach(async () => {
-      await factory.create('User', {
-        email: 'owner@example.org',
-        name: 'Owner',
-        password: 'iamtheowner',
-      })
-      await factory.create('User', {
-        email: 'someone@example.org',
-        name: 'Someone else',
-        password: 'else',
-      })
+      ;[owner, anotherRegularUser, administrator, moderator] = await Promise.all([
+        factory.create('User', {
+          email: 'owner@example.org',
+          name: 'Owner',
+          password: 'iamtheowner',
+        }),
+        factory.create('User', {
+          email: 'another.regular.user@example.org',
+          name: 'Another Regular User',
+          password: 'else',
+        }),
+        factory.create('User', {
+          email: 'admin@example.org',
+          name: 'Admin',
+          password: 'admin',
+          role: 'admin',
+        }),
+        factory.create('User', {
+          email: 'moderator@example.org',
+          name: 'Moderator',
+          password: 'moderator',
+          role: 'moderator',
+        }),
+      ])
+      variables = {}
     })
 
     afterEach(async () => {
       await factory.cleanDatabase()
     })
 
     describe('access email address', () => {
-      let headers = {}
-      let loginCredentials = null
-      const action = async () => {
-        if (loginCredentials) {
-          headers = await login(loginCredentials)
-        }
-        const graphQLClient = new GraphQLClient(host, { headers })
-        return graphQLClient.request('{User(name: "Owner") { email } }')
-      }
-
-      describe('not logged in', () => {
-        it('rejects', async () => {
-          await expect(action()).rejects.toThrow('Not Authorised!')
+      describe('unauthenticated', () => {
+        beforeEach(() => {
+          authenticatedUser = null
         })
-
-        it("does not expose the owner's email address", async () => {
-          let response = {}
-          try {
-            await action()
-          } catch (error) {
-            response = error.response.data
-          } finally {
-            expect(response).toEqual({ User: [null] })
-          }
+        it("throws an error and does not expose the owner's email address", async () => {
+          await expect(
+            query({ query: userQuery, variables: { name: 'Owner' } }),
+          ).resolves.toMatchObject({
+            errors: [{ message: 'Not Authorised!' }],
+            data: { User: [null] },
+          })
```


This is so much better :heart_eyes:

File: webapp/pages/admin/users.vue
Comment (Tirokk) [inline webapp/pages/admin/users.vue:38]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-07T09:00:20Z" title="Monday, October 7th 2019, 11:00:20 am +02:00">Oct 7, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -33,6 +33,11 @@
             <b>{{ scope.row.name | truncate(20) }}</b>
           </nuxt-link>
         </template>
+        <template slot="email" slot-scope="scope">
+          <a :href="`mailto:${scope.row.email}`">
+            <b>{{ scope.row.email }}</b>
```


Some of those emails are quite long. Maybe we should think about how to fit the data into the layout? E.g. at some places in the app, we use `| truncate(20)` filter.

Although I'm OK to merge this and wait for our admins to complain about it :wink:

Comment (Tirokk) [reply webapp/pages/admin/users.vue:38]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-07T09:38:39Z" title="Monday, October 7th 2019, 11:38:39 am +02:00">Oct 7, 2019</time>_

---


That's what I thought as well, @datenbrei mentioned that they might be copying and pasting the links, so maybe it's more difficult if it's truncated??

Comment (Tirokk) [reply webapp/pages/admin/users.vue:38]: > [aonomike](https://github.com/aonomike) **Authored by [aonomike](https://github.com/aonomike)**
_<time datetime="2019-10-07T10:34:45Z" title="Monday, October 7th 2019, 12:34:45 pm +02:00">Oct 7, 2019</time>_

---


Thanks @mattwr18 @roschaefer 
Let me know if I need to update this :D

**Review by Mogge — APPROVED** [review]
Looks good to me @aonomike!
And you even refactored another backend test!!
Great job!!

**Review by Mogge — APPROVED** [review]
hesitant

To me, this looks good and I can't see why we should not merge it. It's tested and there is nothing that comes to my mind, which we need to do before merging. Great job @aonomike !

But I'm a little hesitant to merge this before our daily standup. We could hear back from the others if we missed anything important here. it's a matter of security/data-privacy after all :wink:

**Pull request closed without merging**
